Add CMake support

ExoSkye opened the pull request:
This PR adds support for CMake as a build system
This approach is heavily based on nxdk's CMake system (https://github.com/XboxDev/nxdk/pull/518)
This is also a better version of #125

ExoSkye committed:
Add preliminary CMake support

Things left to do:
- Add PPU support
- Add support for building shaders
- Testing

ProtoByter commented:
This approach is heavily based on nxdk's CMake system

ExoSkye committed:
Fixed pkg support

Also made a couple of code changes to make the code more readable and easier to use
For example:
- It now doesn't use one single target for generating all the pkgs a project could make
- It doesn't use `PSL1GHT` as a variable since most of the pkg code was copied from the old version and used that for the path to `psl1ght` whilst for this version it uses `PS3DEV_DIR`

This has also been tested more so that's good

ProtoByter commented:
Things left to do:
- Support SPU code
- Shader building code
- Add something equivalent to the `data_rules` makefile
- Maybe port some of the examples?

ProtoByter commented:
The SPU code will be tricky, since you can't use two compilers for one language in one CMake project so I have a feeling I might need to resort to having two projects, one with the SPU code and one with the PPU code and the PPU code project has a `add_custom_command` call to tell CMake to build the SPU project, then use the binaries that generates

Although I don't really like that since it feels messy

ExoSkye committed:
Made a couple of QoL improvements

Added comments in code and in the custom commands to say what steps/functions do what
Added shader compilation custom commands
Fixed copying the content into the staging directory
Fixed building on windows by including `TOOL_OS_SUFFIX`

ProtoByter commented:
So I've added shader building code, although it isn't tested yet

miigotu commented:
Maybe the share folder should be called contrib? Also, can you add a a sample project back that uses this build method and runs on ps?

ProtoByter commented:
Oh, yea, I was just basing the name (`share`) off of what it is in `nxdk`, I'll change that when I'm on my PC
And yeah, I'll also make a sample project demonstrating how to do stuff on the PPU. I'll also make an SPU example when I add SPU support

ProtoByter commented:
Also what standard libraries are there in psl1ght? I've set it as `libc`, `liblv2` and `librt` currently but are there any more that'd be considered standard?

ExoSkye committed:
Move to using `contrib` directory and add template projects

ProtoByter commented:
> Maybe the share folder should be called contrib? Also, can you add a a sample project back that uses this build method and runs on ps? An spu and a ppu example would be nice.

Ok, I've added the example projects and moved it now

miigotu commented:
I'm not sure what stdlibs we have set up, it's been a LONG time 🤣. But this looks good to me, and any additions or adjustments that may be needed they can always be made later.

ExoSkye committed:
Make a CMake SPU Sample

This is untested, so I don't fully know if it works on a PS3, I just know it builds

ExoSkye committed:
Add a utility to automatically create an sfo.xml from some information

ExoSkye committed:
Add a utility to automatically create an sfo.xml from some information

ProtoByter commented:
The SPU Sample that I've added is untested and is based on the `sputest` example
The sfo.xml stuff just seemed like it'd be useful

ExoSkye committed:
Merge remote-tracking branch 'ProtoByter/cmake' into cmake

ProtoByter commented:
I had to amend it because I forgot to update the SPU sample to use the `sfo.xml` generation features

ExoSkye committed:
Switch versions to the correct format and don't use VERSION for APP_VER

`APP_VER` just has a default of `01.00` instead of being the same as `VERSION`

ExoSkye committed:
Fix icon copying

ExoSkye committed:
Fix content copying

ExoSkye committed:
Add `data_rules` equivalent

ProtoByter commented:
How does this look now? @miigotu

miigotu commented on contrib/ppu-helpers.cmake:
```diff
@@ -0,0 +1,111 @@
+    )
+endfunction()
+
+function(add_spu_project path target_name)
```
It does, thanks.

ExoSkye committed:
Change `add_bin_file`'s comment to be more readable

miigotu commented:
I'll have to set this up and build something on my machine later. I'll let you know how it goes. I think in another PR we could extract some of the variables that are set and put them in a separate cmake definition file and include them. A lot of toolchain variables are set that normally would be invisible to the end user/developer that I think could be pulled away and make the front facing cmake file easier to read. Many of those variables are the same for spu and ppu.

ProtoByter commented:
What variables are you talking about?

miigotu commented:
> Which variables are you talking about?

The shared code between spu-toolchain.cmake and ppu-toochain.cmake

ProtoByter commented:
Oh ic, I think it should be possible to make a `base-tooolchain` file and then the other two files set some variables before including it

ProtoByter commented:
Also, it appears that the executables outputted appear to not be able to run in rpcs3, and I haven't ported the `blitting` example to CMake so I can't test it on an real PS3 since it doesn't do anything on the screen (so I can't see if it works)
It appears to get stuck around the area where it's initializing syscalls which is weird. So I'm currently trying to debug that

miigotu commented:
Fself won't work without a patched OS. Also, there are some bugs in geohots tools that I don't think are fixed here yet, many people used the scetoool. I don't have a PS3 anymore (like yearsss) so I can't help there lol.

ProtoByter commented:
Oh I think that's got nothing to do with fself
it appears to quit during `__do_global_ctors_aux`

miigotu commented:
> Oh I think that's got nothing to do with fself
> it appears to quit during `__do_global_ctors_aux`

That is in the constructor of the elf, which could be trying to verify or access some syscall. I'm not sure which elf you are trying to use or if rpcs3 allows fake signed elf's, or patches lv2 or verification.

miigotu commented:
It's actually in the constructor of a static object ^ in C++

ProtoByter commented:
Oh I see, I'll try to see if a C executable works then
Also, any fself made by the makefile system works in rpcs3 (not the elfs though)

ProtoByter commented:
C executable still fails there :/

ProtoByter commented:
So it appears that to create a new lock (used in `glob_init`) it first has to lock a `metaLock`. This `metaLock`, however, is just null bytes so locking it fails. This is really weird

ProtoByter commented:
So I found the issue, this is the `.ctors` section that the executable made using this gets:
![image](https://user-images.githubusercontent.com/20995169/162568656-ad96dddb-84f7-4e53-8e76-61b818ff557a.png)
Whilst the PSL1GHT Makefiles give this:
![image](https://user-images.githubusercontent.com/20995169/162568673-aea6d45a-b504-457b-8115-f95035bf6a7f.png)
As you can see, `glob_file_init` needs to be called after `init_metalock`. However I'm not too sure why the compiler does this and what is different about what psl1ght is doing compared to CMake

miigotu commented:
Is the sbrk_init supposed to use the same ptr between both executables?

ProtoByter commented:
Yes, since these were both generated by the same linker, I believe they should be the same

miigotu commented:
I agree. So I think it's not using the same compiler or linker somewhere. Or possibly, not getting the right header included (eg: a system header instead of the toolchain)

ProtoByter commented:
With the help of a friend, I managed to find the issue, the `metaLock` init function had the same priority than the `glob_file` init function, this makes no sense since `glob_file` init function relies on the `metaLock` so depending on link order, the `glob_file` init function could be run before `metaLock` was initialized

ProtoByter commented:
However, for whatever reason stdio doesn't get initialized, which is strange, so printf doesn't eork

ProtoByter commented:
Wait, why isn't `libc.c` from `librt` being built with `librt`? Since that has the initialization function for `libc`

ProtoByter commented:
Yea

miigotu commented:
I don't really know where to help, but maybe it's a toolchain header getting included somewhere instead of psl1ght. I'd check libdirs and include paths in the cmake.

NotNite commented:
Sorry to necropost, but was this issue with the locks ever figured out? I'm trying to get PSL1GHT working in Meson and running into what seems to be a similar issue.

ExoSkye commented:
> Sorry to necropost, but was this issue with the locks ever figured out? I'm trying to get PSL1GHT working in Meson and running into what seems to be a similar issue.

I did not, sadly

williamkibira commented:
@ExoSkye, thank you for making this template. It's a little less intimidating compared to the Makefile version.
I've managed to derive something from to be able to compile for spu based projects.

Going to try making a simple home brew project and see how everything goes.

Fancy2209 commented:
Hello, I've been using CMake for PPU elf's for some weeks now with no problems, (as long as I remeber sprxlinker should be ran on the elf), here's my CMake file, I just run sprxlinker on the elf manually after, hopefully it's of use for reviving this?
https://raw.githubusercontent.com/Fancy2209/mgba/refs/heads/master/src/platform/ps3/CMakeToolchain.txt
